presto-main/src/main/java/com/facebook/presto/execution/resourceGroups/InternalResourceGroup.java

```diff
@@ -710,18 +709,18 @@ private void startInBackground(QueryExecution query)
         }
     }
 
-    private void queryFinished(QueryExecution query)
+    private void queryFinished(ManagedQueryExecution query)
     {
         synchronized (root) {
             if (!runningQueries.contains(query) && !queuedQueries.contains(query)) {
                 // Query has already been cleaned up
                 return;
             }
             // Only count the CPU time if the query succeeded, or the failure was the fault of the user
-            if (query.getState() == QueryState.FINISHED || query.getQueryInfo().getErrorType() == USER_ERROR) {
+            if (query.getErrorCode().isPresent() || query.getErrorCode().get().getType() == USER_ERROR) {
```

Comment:
To have this form, you need to prove that `getErrorCode` is only present when the query is in the failed state.  Also, can you document this in the interface.